csharp/src/Google.Protobuf/MessageExtensions.cs

```diff
@@ -140,6 +143,75 @@ public static ByteString ToByteString(this IMessage message)
             return ByteString.AttachBytes(message.ToByteArray());
         }
 
+        /// <summary>
+        /// Checks if all required fields in a message have values set. For proto3 messages, this returns true
+        /// </summary>
+        public static bool IsInitialized(this IMessage message)
+        {
+            if (message.Descriptor.File.Proto.Syntax != "proto2")
+            {
+                return true;
+            }
+
+            return message.Descriptor
+                .Fields
+                .InDeclarationOrder()
+                .Where(f => f.IsRequired || f.IsRepeated || f.IsMap || f.MessageType != null)
+                .All(f =>
+                {
+                    if (f.MessageType != null)
+                    {
+                        if (f.Accessor.HasValue(message))
+                        {
+                            var field = (IMessage)f.Accessor.GetValue(message);
+                            return field.IsInitialized();
+                        }
+                        else
+                        {
+                            return false;
+                        }
+                    }
+                    else if (f.IsRepeated)
+                    {
+                        IList list = (IList)f.Accessor.GetValue(message);
+                        foreach (IMessage item in list)
+                        {
+                            if (item is null)
+                            {
+                                return true;
```

Comment:
It will stop the foreach (IMessage item in list) loop when one item is true?

Reply:
Yes, if the item in the list is not an IMessage, the `item` will be null. Since you can't add null values to repeated fields, this doesn't risk returning early for null messages

Reply:
Do you mean when it is a repeated scalar field? I'd prefer to check the field descriptor before the for loop?